torch/csrc/jit/python/init.cpp

```diff
@@ -165,7 +165,9 @@ void initJITBindings(PyObject* module) {
   auto m = py::handle(module).cast<py::module>();
   auto jit = m.def_submodule("_jit");
 
-  static py::exception<JITException> exc(m, "JITException");
+  // This is a static object, so we must leak the Python object
+  static py::handle exc =
+      py::exception<JITException>(m, "JITException").release();
```

Comment:
I'm not very familiar with PyBind, but if object is allocated that way, what would prevent Python runtime from grabage collecting it?
I know this looks ugly, but if we want to keep the object referenced but never call it'd destructor perhaps something like `static auto * exc = new py::exception<JITException>(m, "JITException");` will do better? (Or I think I have a dummy unique_pointer with leaky destrutor)

Reply:
That might actually be better!

(I guess I was too trapped in the `handle` idea which happens to work for our purposes just fine, I didn't really consider alternatives.)

Reply:
> what would prevent Python runtime from grabage collecting it

The reference count will never go to zero, the deallocation will never trigger.

Python doesn't keep a registry of all objects AFAIK, at least not in production configuration, therefore I think the `Py_DECREF` trigger is the only way there is to get to the deallocation code. Please correct me if I'm wrong.

Reply:
> something like static auto * exc = new py::exception<JITException>(m, "JITException"); will do better?

@malfet @albanD Could you please take a look here?

https://github.com/pybind/pybind11/pull/4772/files

(For future reference, at the point of this writing that was https://github.com/pybind/pybind11/pull/4772/commits/9a4eba52293f4ca60d0cbd13253f506ffe8d5413)

I'm having 2nd thoughts, although it's 99.99% a style opinion. (0.01% that there is an extra dynamic memory allocation with the `new` alternative that is not actually needed.)

My thinking:

* Using the `py::handle ... .release();` alternative combined with `py::set_error()` is more concise / looks nicer IMO, seems very intuitive to me, compared to the surprising and evidently troublesome `operator()` overload.

* In retrospect, we should have added `py::set_error()` a very long time ago. (There are 42 `PyErr_SetString()` across all files in the pybind11 repo, I've not replaced them all yet.)

Please let me know what you think.

Reply:
> Python doesn't keep a registry of all objects AFAIK

It does keep such a registry for all GC-able objects.
But release() leak one reference indeed. So this objects refcount will never be able to reach 0 and so will never be able to be de-allocated.

> Could you please take a look here?

I agree that this is very much style. My personal vote is that avoiding the `new` is simpler in this context.
@malfet are you ok with me merging the current version of the PR? (I'll update to py::set_error() once we upgrade to a version of pybind that has that).

Reply:
Sounds fine to me, build please add a comment saying that this object will never be GCed because of the reasons outlined above